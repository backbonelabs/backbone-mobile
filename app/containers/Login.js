import React, { Component } from 'react';
import {
  ActivityIndicator,
  Alert,
  View,
  Text,
  TextInput,
  TouchableHighlight,
} from 'react-native';
import { connect } from 'react-redux';
import SensitiveInfo from '../utils/SensitiveInfo';
import authActions from '../actions/auth';
import styles from '../styles/login';
import routes from '../routes';

class Login extends Component {
  static propTypes = {
    accessToken: React.PropTypes.string,
    errorMessage: React.PropTypes.string,
    dispatch: React.PropTypes.func,
    isFetching: React.PropTypes.bool,
    navigator: React.PropTypes.object,
  };

  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
    };
    this.login = this.login.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (!this.props.accessToken && nextProps.accessToken) {
      // User successfully authenticated, save access token to local device
      this.saveAccessToken(nextProps.accessToken);

      // Redirect to Home
      this.props.navigator.replace(routes.home);
    } else if (!this.props.errorMessage && nextProps.errorMessage) {
      // Authentication error
      Alert.alert('Authentication Error', 'Invalid email/password. Please try again.');
    }
  }

  saveAccessToken(accessToken) {
    SensitiveInfo.setItem('accessToken', accessToken);
  }

  login() {
    if (!this.state.email || !this.state.password) {
      // Show alert if email or password is missing
      Alert.alert('Missing fields', `${this.state.email ? 'Password' : 'Email'} is required`);
    } else {
      this.setState({ submitted: true }, () => {
        const { email, password } = this.state;
        this.props.dispatch(authActions.login({ email, password }));
      });
    }
  }

  render() {
    // TODO: Show ActivityIndicator when form is submitted
    return (
      <View style={styles.container}>
      {this.props.isFetching ?
        <ActivityIndicator
          animating
          size="large"
          color={styles._activityIndicator.color}
        />
        :
        <View style={styles.formContainer}>
          <View style={styles.textFieldView}>
            <TextInput
              ref={ref => {
                this.emailField = ref;
              }}
              style={styles.textField}
              value={this.state.email}
              autoCapitalize="none"
              placeholder="Email"
              keyboardType="email-address"
              onChangeText={text => this.setState({ email: text })}
              onSubmitEditing={() => this.passwordField.focus()}
              autoCorrect={false}
              autoFocus
              returnKeyType="next"
            />
          </View>
          <View style={styles.textFieldView}>
            <TextInput
              ref={ref => {
                this.passwordField = ref;
              }}
              style={styles.textField}
              value={this.state.password}
              autoCapitalize="none"
              placeholder="Password"
              keyboardType="default"
              onChangeText={text => this.setState({ password: text })}
              onSubmitEditing={this.login}
              autoCorrect={false}
              secureTextEntry
              returnKeyType="go"
            />
          </View>
          <TouchableHighlight
            style={styles.button}
            disabled={this.props.isFetching}
            onPress={this.login}
          >
            <Text style={styles.buttonText}>Log in</Text>
          </TouchableHighlight>
        </View>
      }
      </View>
    );
  }
}

const mapStateToProps = state => {
  const { auth } = state;
  return auth;
};

export default connect(mapStateToProps)(Login);
