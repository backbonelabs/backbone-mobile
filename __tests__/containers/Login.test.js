import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { asyncActionMiddleware } from 'redux-async-action';
import Login from '../../app/containers/Login';

describe('Login Component', () => {
  const initialState = {
    auth: {},
    user: {},
  };
  const store = configureStore([asyncActionMiddleware, thunk])(initialState);
  const wrapper = shallow(
    <Login />,
    { context: { store } },
  );
  let render;

  beforeEach(() => {
    store.clearActions();
    render = wrapper.dive();
  });

  test('renders as expected', () => {
    expect(render).toMatchSnapshot();
  });

  test('Login action dispatched when clicked', () => {
    const btn = render.find('Button').at(0);
    render.setState({ email: 'testing@mail.com', password: 'password' });
    btn.simulate('press');
    const actions = store.getActions();
    expect(actions).toEqual([{ type: 'LOGIN__START' }]);
  });

  test('Login button disabled on first render', () => {
    const btn = render.find('Button').at(0);
    expect(btn.props().disabled).toEqual(true);
  });

  test('Login button disabled remove', () => {
    render.setState({ email: 'testing@mail.com', password: 'password' });
    const btn = render.find('Button').at(0);
    expect(btn.props().disabled).toEqual(false);
  });

  test('Email state updates when text change', () => {
    const emailInput = render.find('Input').at(0);
    const nextValue = 'some email';
    emailInput.simulate('changeText', nextValue);
    expect(render.state().email).toEqual(nextValue);
  });

  test('Password state updates when text change', () => {
    const passwordInput = render.find('Input').at(1);
    const nextValue = 'some password';
    passwordInput.simulate('changeText', nextValue);
    expect(render.state().password).toEqual(nextValue);
  });
});

