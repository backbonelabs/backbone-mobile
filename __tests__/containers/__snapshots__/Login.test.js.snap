// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component renders as expected 1`] = `
<TouchableWithoutFeedback
  onPress={[Function]}
>
  <View
    style={
      Array [
        undefined,
        Object {
          "height": 1314,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            undefined,
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View>
          <Image
            source={1}
            style={
              Array [
                undefined,
                Object {
                  "height": 220,
                },
              ]
            }
          />
        </View>
        <View>
          <TouchableOpacity
            activeOpacity={0.2}
            focusedOpacity={0.7}
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            <BodyText>
              Log In
            </BodyText>
          </TouchableOpacity>
          <TouchableOpacity
            activeOpacity={0.2}
            focusedOpacity={0.7}
            onPress={[Function]}
          >
            <BodyText>
              Sign Up
            </BodyText>
          </TouchableOpacity>
        </View>
      </View>
      <View>
        <View>
          <View>
            <Connect(Facebook)
              buttonText="LOG IN WITH FACEBOOK"
            />
            <View>
              <View />
              <SecondaryText>
                OR
              </SecondaryText>
              <View />
            </View>
          </View>
          <Input
            autoCapitalize="none"
            autoCorrect={false}
            handleRef={[Function]}
            iconFont="MaterialIcon"
            iconLeftName="email"
            iconStyle={
              Object {
                "color": "#9E9E9E",
              }
            }
            keyboardType="email-address"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Email"
            returnKeyType="next"
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
              }
            }
            value=""
          />
          <Input
            autoCapitalize="none"
            autoCorrect={false}
            handleRef={[Function]}
            iconFont="MaterialIcon"
            iconLeftName="lock"
            iconStyle={
              Object {
                "color": "#9E9E9E",
              }
            }
            keyboardType="default"
            onChangeText={[Function]}
            onSubmitEditing={null}
            placeholder="Password"
            returnKeyType="go"
            secureTextEntry={true}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
              }
            }
            value=""
          />
          <TouchableOpacity
            activeOpacity={0.4}
            focusedOpacity={0.7}
            onPress={[Function]}
          >
            <SecondaryText>
              Forgot your password?
            </SecondaryText>
          </TouchableOpacity>
        </View>
        <View>
          <Button
            disabled={true}
            onPress={[Function]}
            primary={true}
            shadow={true}
            style={Object {}}
            text="LOG IN"
            textStyle={Object {}}
          />
        </View>
      </View>
    </View>
  </View>
</TouchableWithoutFeedback>
`;
